src/main/java/edu/kit/kastel/sdq/artemis4j/grading/Assessment.java

```diff
@@ -0,0 +1,427 @@
+     */
+    public void addPredefinedAnnotation(MistakeType mistakeType, String filePath, int startLine, int endLine, String customMessage) {
+        if (mistakeType.isCustomAnnotation()) {
+            throw new IllegalArgumentException("Mistake type is a custom annotation");
+        }
+
+        var source = this.correctionRound == 0 ? Annotation.AnnotationSource.MANUAL_FIRST_ROUND : Annotation.AnnotationSource.MANUAL_SECOND_ROUND;
+        this.annotations.add(new Annotation(mistakeType, filePath, startLine, endLine, customMessage, null, source));
+    }
+
+    /**
+     * Adds a custom manual annotation to the assessment.
+     * @param mistakeType Must be a custom mistake type
+     * @param filePath
+     * @param startLine
+     * @param endLine
+     * @param customMessage May not be null
+     * @param customScore
+     */
+    public void addCustomAnnotation(MistakeType mistakeType, String filePath, int startLine, int endLine, String customMessage, double customScore) {
+        if (!mistakeType.isCustomAnnotation()) {
+            throw new IllegalArgumentException("Mistake type is not a custom annotation");
+        }
+
+        var source = this.correctionRound == 0 ? Annotation.AnnotationSource.MANUAL_FIRST_ROUND : Annotation.AnnotationSource.MANUAL_SECOND_ROUND;
+        this.annotations.add(new Annotation(mistakeType, filePath, startLine, endLine, customMessage, customScore, source));
+    }
+
+    public void addAutograderAnnotation(MistakeType mistakeType, String filePath, int startLine, int endLine, String explanation) {
+        var customScore = mistakeType.isCustomAnnotation() ? 0.0 : null;
+        this.annotations.add(new Annotation(mistakeType, filePath, startLine, endLine, explanation, customScore, Annotation.AnnotationSource.AUTOGRADER));
+    }
+
+    /**
+     * Removes an annotation from the assessment. If the annotation is not present, nothing happens.
+     *
+     * @param annotation
+     */
+    public void removeAnnotation(Annotation annotation) {
+        this.annotations.remove(annotation);
+    }
+
+    /**
+     * Clears all annotations from the assessment.
+     */
+    public void clearAnnotations() {
+        this.annotations.clear();
+    }
+
+    private void internalSaveOrSubmit(boolean shouldSubmit, Locale locale) throws AnnotationMappingException, ArtemisNetworkException {
+        log.info("Packing assessment for artemis");
+        var feedbacks = this.packAssessmentForArtemis(locale);
+        var absoluteScore = this.calculateTotalPoints();
+        var relativeScore = absoluteScore / this.getMaxPoints() * 100.0;
+        var result = ResultDTO.forAssessmentSubmission(this.programmingSubmission.getId(), relativeScore, feedbacks, this.getConnection().getAssessor().toDTO());
+
+        // Sanity check
+        var feedbackPoints = result.feedbacks().stream().mapToDouble(FeedbackDTO::credits).sum();
+        if (absoluteScore != feedbackPoints) {
```

Comment:
Please avoid var for any primitive type. Its hard to review if I don't know the types